components/hab/src/command/pkg/binlink.rs

```diff
@@ -135,89 +145,165 @@ pub fn binlink_all_in_pkg(ui: &mut UI,
     Ok(())
 }
 
-fn is_dest_on_path<T: AsRef<Path>>(dest_dir: T) -> bool {
+fn is_dest_on_path(dest_dir: &Path) -> bool {
     if let Some(val) = env::var_os("PATH") {
-        env::split_paths(&val).any(|p| p == dest_dir.as_ref())
+        env::split_paths(&val).any(|p| p == dest_dir)
     } else {
         false
     }
 }
 
+#[allow(dead_code)]
 struct Binlink {
-    dest: PathBuf,
-    src:  PathBuf,
+    link_path:           PathBuf,
+    target_path:         PathBuf,
+    default_interpreter: PathBuf,
 }
 
-#[cfg(not(target_os = "windows"))]
 impl Binlink {
-    pub fn new<T: AsRef<Path>>(src: T, dest_dir: T) -> Result<Self> {
-        let bin_name = match src.as_ref().file_name() {
-            Some(name) => name,
-            None => return Err(Error::CannotParseBinlinkSource(src.as_ref().to_path_buf())),
-        };
-
-        Ok(Self { dest: dest_dir.as_ref().join(bin_name),
-                  src:  src.as_ref().to_path_buf(), })
+    pub fn new(target_path: &Path, link_path: &Path, default_interpreter: &Path) -> Result<Self> {
+        Ok(Self { link_path:           Self::binstub_path(&target_path, link_path)?,
+                  target_path:         target_path.to_path_buf(),
+                  default_interpreter: default_interpreter.to_path_buf(), })
     }
 
-    pub fn from_file<T: AsRef<Path>>(dest: T) -> Result<Self> {
-        Ok(Binlink { dest: dest.as_ref().to_path_buf(),
-                     src:  fs::read_link(&dest)?, })
+    pub fn from_file(path: &Path) -> Result<Self> {
+        // its possible the link could already exist as a symlink either
+        // because a previous version of habitat did the binlinking or
+        // the symlink created outside of habitat
+        fs::read_link(path).map(|target_path| Self::from_link(target_path, path.into()))
+                           .or_else(|_| Self::try_from_script(path))
     }
 
-    pub fn link(&self) -> Result<()> {
-        use crate::hcore::os::filesystem;
+    fn from_link(target_path: PathBuf, link_path: PathBuf) -> Self {
+        Self { link_path,
+               target_path,
+               default_interpreter: PathBuf::from(DEFAULT_INTERPRETER) }
+    }
 
-        filesystem::symlink(&self.src, &self.dest)?;
-        Ok(())
+    fn try_from_script(path: &Path) -> Result<Self> {
```

Comment:
Maybe call this `script` or `binstub`? I think it makes things clearer later when there are multiple paths being juggled around.